File: branches/SUSD/generic_scene_graph/GSG/Core/Node.cpp

///////////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2002, Perry L. Miller IV
//  All rights reserved.
//  BSD License: http://www.opensource.org/licenses/bsd-license.html
//
///////////////////////////////////////////////////////////////////////////////

///////////////////////////////////////////////////////////////////////////////
//
//  Base class for all scene graph nodes.
//
///////////////////////////////////////////////////////////////////////////////

#include "GSG/Core/Precompiled.h"
#include "GSG/Core/Node.h"
#include "GSG/Core/Visitor.h"
#include "GSG/Core/PushPop.h"

using namespace GSG;

GSG_IMPLEMENT_ACCEPT_NODE ( Node );


/////////////////////////////////////////////////////////////////////////////
//
//  Constructor.
//
/////////////////////////////////////////////////////////////////////////////

Node::Node() : Referenced(),
  _bound()
{
}


/////////////////////////////////////////////////////////////////////////////
//
//  Copy constructor.
//
/////////////////////////////////////////////////////////////////////////////

Node::Node ( const Node &n ) : Referenced ( n ),
  _bound ( n._bound )
{
}


/////////////////////////////////////////////////////////////////////////////
//
//  Destructor.
//
/////////////////////////////////////////////////////////////////////////////

Node::~Node()
{
}


/////////////////////////////////////////////////////////////////////////////
//
//  Update this node.
//
/////////////////////////////////////////////////////////////////////////////

void Node::update()
{
  Lock lock ( this );

  // Calculate the bounding sphere.
  this->calculateBoundingSphere();
}


/////////////////////////////////////////////////////////////////////////////
//
//  Get the bounding sphere.
//
/////////////////////////////////////////////////////////////////////////////

const BoundingSphere &Node::boundingSphere() const
{
  Lock lock ( this );
  return _bound;
}


/////////////////////////////////////////////////////////////////////////////
//
//  Set the bounding sphere.
//
/////////////////////////////////////////////////////////////////////////////

void Node::boundingSphere ( const BoundingSphere &b )
{
  Lock lock ( this );
  _bound = b;
}

File: branches/usul_mutex_test/Legacy/DotNet/Scene/Nodes/Node.cs

///////////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2006, Perry L Miller IV
//  All rights reserved.
//  BSD License: http://www.opensource.org/licenses/bsd-license.html
//
///////////////////////////////////////////////////////////////////////////////

namespace CadKit.Scene.Nodes
{
  public class Node : CadKit.Scene.Base.Object
  {
    /// <summary>
    /// Data members.
    /// </summary>
    private object _data = null;
    private string _name = null;

    /// <summary>
    /// Delegates
    /// </summary>
    public delegate void VoidReturnWithNoArguments();

    /// <summary>
    /// Constructor
    /// </summary>
    protected Node()
      : base()
    {
    }

    /// <summary>
    /// Destructor
    /// </summary>
    ~Node()
    {
    }

    /// <summary>
    /// Clean this instance.
    /// </summary>
    protected override void _cleanup()
    {
      lock (this.Mutex)
      {
        base._cleanup();
      }
    }

    /// <summary>
    /// Accept the visitor
    /// </summary>
    public virtual void accept(CadKit.Scene.Visitors.Visitor visitor)
    {
      visitor.visit(this);
    }

    /// <summary>
    /// Set/get the user data.
    /// </summary>
    public object UserData
    {
      get { lock (this.Mutex) { return _data; } }
      set { lock (this.Mutex) { _data = value; } }
    }

    /// <summary>
    /// Get/set the name.
    /// </summary>
    public string Name
    {
      // Return a copy.
      get { lock (this.Mutex) { return ((null == _name) ? null : System.String.Copy(_name)); } }
      set { lock (this.Mutex) { _name = value; } }
    }
  }
}
